Sources/Kafka/RDKafka/RDKafkaClient.swift

```diff
@@ -451,18 +411,30 @@ final class RDKafkaClient: Sendable {
         actualCallback(.success(()))
     }
 
-    /// Redirect the main ``RDKafkaClient/poll(timeout:)`` queue to the `KafkaConsumer`'s
-    /// queue (``RDKafkaClient/consumerPoll``).
+    /// Request a new message from the Kafka cluster.
     ///
-    /// Events that would be triggered by ``RDKafkaClient/poll(timeout:)``
-    /// are now triggered by ``RDKafkaClient/consumerPoll``.
+    /// - Important: This method should only be invoked from ``KafkaConsumer``.
     ///
-    /// - Warning: It is not allowed to call ``RDKafkaClient/poll(timeout:)`` after ``RDKafkaClient/pollSetConsumer``.
-    func pollSetConsumer() throws {
-        let result = rd_kafka_poll_set_consumer(self.kafkaHandle)
-        if result != RD_KAFKA_RESP_ERR_NO_ERROR {
-            throw KafkaError.rdKafkaError(wrapping: result)
+    /// - Returns: A ``KafkaConsumerMessage`` or `nil` if there are no new messages.
+    /// - Throws: A ``KafkaError`` if the received message is an error message or malformed.
+    func consumerPoll() throws -> KafkaConsumerMessage? {
+        guard let messagePointer = rd_kafka_consumer_poll(self.kafkaHandle, 0) else {
+            // No error, there might be no more messages
+            return nil
         }
+
+        defer {
+            // Destroy message otherwise poll() will block forever
+            rd_kafka_message_destroy(messagePointer)
+        }
+
+        // Reached the end of the topic+partition queue on the broker
+        if messagePointer.pointee.err == RD_KAFKA_RESP_ERR__PARTITION_EOF {
```

Comment:
Hm, I wonder if now all errors will come here or some of them will be received in `eventsPoll()`

Reply:
I think there is going to be a clear split now between message related errors and run loop errors.